services/mailer/incoming/incoming_handler.go

```diff
@@ -71,11 +71,16 @@ func (h *ReplyHandler) Handle(ctx context.Context, content *MailContent, doer *u
 		return err
 	}
 
-	if !perm.CanWriteIssuesOrPulls(issue.IsPull) || issue.IsLocked && !doer.IsAdmin {
+	if !perm.CanWriteIssuesOrPulls(issue.IsPull) && issue.IsLocked && !doer.IsAdmin {
```

Comment:
I think the old code tried to do something like:

- Return nil if the doer cannot write
- Even if the doer can write, return nil if the issue is locked and the doer is not admin.

So maybe we could just change it to:

```go
if !perm.CanReadIssuesOrPulls(issue.IsPull) || issue.IsLocked && !doer.IsAdmin {
```

Reply:
but in the api, user with write permission can comment on locked issue. ref:
https://github.com/go-gitea/gitea/blob/9f919cf08339e8ec8f6b0b2a0dcafb3ea905fbec/routers/api/v1/repo/issue_comment.go#L360-L363

if use your logic, they can't. I wonder ...

Reply:
I see, could you please change it to `if issue.IsLocked && ...` too? It does make it easier to understand.

Reply:
```suggestion
	if issue.IsLocked && !perm.CanWriteIssuesOrPulls(issue.IsPull) && !doer.IsAdmin {
```

Reply:
Yes, so it will be clear that this is to deal with locked issues.

Reply:
```suggestion
	// Locked issues require write permissions
	if issue.IsLocked && !perm.CanWriteIssuesOrPulls(issue.IsPull) && !doer.IsAdmin {
```

Reply:
Wait, is `Read` really enough to `Write` an issue comment?
I can see a lot of problems waiting for us, i.e. when we add private issues…

Reply:
As I said: That will certainly make implementing private issues (safely) a pain.

Reply:
> As I said: That will certainly make implementing private issues (safely) a pain.

I think has been diffcult now ... , when implementing private issues. it's better to use a unique servcie fuction to create commnt for both web ui, api and incoming mail...